### DISCUSSION TITLE
Support string array for devServer.static.directory

### DISCUSSION THREAD

**Discussion by PierBJX** [webpack/webpack#15076] [Ideas]
Using webpack dev server, it is possible to serve static files using the property:
```
module.exports = {
  // ...
  devServer: {
    static: [path.join(__dirname, 'assets'), path.join(__dirname, 'css')],
  },
};
```

It works properly, however, if I want to add a `publicPath`, I need to create two objects like this:
```
module.exports = {
  // ...
  devServer: {
    static: [
       {
          directory: path.join(__dirname, 'assets'),
          publicPath: '/serve-public-path-url',
       },
       {
          directory: path.join(__dirname, 'css'),
          publicPath: '/serve-public-path-url',
       },
    ],
  },
};
```

Whereas, it could make sense to create only one object with two paths in the directory such like that:
```
module.exports = {
  // ...
  devServer: {
    static:  {
        directory: [path.join(__dirname, 'assets'), path.join(__dirname, 'css')],
        publicPath: '/serve-public-path-url',
    }
  },
};
```

What do you think ?

**Comment by alexander-akait** [comment]
Yes, because `static` middleware accepts only on directory https://github.com/expressjs/serve-static, but we can handle them using multiple public paths, it was done to avoid misleading and complex configurations which can lead to the problems

**Reply by PierBJX** [reply]
Alright, so it is not achievable?

**Reply by alexander-akait** [reply]
In theory we can improve it, but I recommend to use multiple objects, like you do right now

**Reply by PierBJX** [reply]
Alright I see :)

By the way, about types, the property `devServer` is not available in `Configuration` (alias for `WebpackOptions`) but only in `WebpackOptionsNormalized`. Do you know why?

**Reply by alexander-akait** [reply]
Please update dev server to the latest version, recently we added types and do it https://github.com/webpack/webpack-dev-server/blob/master/types/lib/Server.d.ts#L3379, so `devServer` should be typed

**Reply by PierBJX** [reply]
So I should be able to do?

```
import { Configuration } from 'webpack';

const devServerOptions: Configuration['devServer'] = {
  ...
}
```

**Reply by alexander-akait** [reply]
Yes

**Reply by PierBJX** [reply]
I have this error:
> TS2339: Property 'devServer' does not exist on type 'Configuration'.

These are my versions:
```
    "webpack": "^5.65.0",
    "webpack-cli": "^4.9.1",
    "webpack-dev-server": "^4.7.2",
    "webpack-merge": "^5.8.0"
```

**Reply by alexander-akait** [reply]
Weird, we have https://github.com/webpack/webpack-dev-server/blob/master/package.json#L7 and https://github.com/webpack/webpack-dev-server/blob/master/types/lib/Server.d.ts#L3380, locally works fine too

**Reply by alexander-akait** [reply]
Oh, I see, give me time to investigate

**Reply by alexander-akait** [reply]
Please use
```ts
/// <reference path="node_modules/webpack-dev-server/types/lib/Server.d.ts"/>
import type { Configuration } from "webpack";

// Your logic
```

Or you can import type from `webpack-dev-server`, i.e.
```ts
import type { Configuration as DevServerConfiguration } from "webpack-dev-server";
import type { Configuration } from "webpack";

const devServer: DevServerConfiguration = {};
const config: Configuration = { devServer };

// module.exports
export default config;
```

typescript doesn't import other types from other packges by default, so we need to say it (also it can be done using `tsconfig.json`)

**Reply by PierBJX** [reply]
Thanks!

**Reply by evenfrost** [reply]
@alexander-akait would be great if this could be added to the readme. This is a pretty common error when you're adding `webpack-dev-server` in your TypeScript webpack project, and googling first leads to the `@types/webpack-dev-server` issue, which is not relevant anymore.

**Reply by alexander-akait** [reply]
@evenfrost Yes, make sense, can you help us and send a PR? It should be easy

**Reply by evenfrost** [reply]
@alexander-akait [done](https://github.com/webpack/webpack-dev-server/pull/4301).
